### DISCUSSION TITLE
How do you check if a connection is open?

### DISCUSSION THREAD

**Discussion by TheWicklowWolf** [FreeOpcUa/opcua-asyncio#1230] [Q&A] [answered]
I open a connection using:
```
from asyncua.sync import Client
conn = Client("serveraddress")
conn.connect()
```
before calling `conn.disconnect()` I would like to check if the connection is open.
Thanks

**Comment by ZuZuD** [comment]
Just as an aside, your username disturbs my productivity a lot :) When I receive a notification from TheWicklowWolf, I always think (hope?) I'll have receive some nice free beers! Are you from the county?

Thus said, you can check if the client is connected with something like this:
```
c = asyncua.Client(url="opc.tcp://127.0.0.1:4840")
c.uaclient.protocol == None 
Out[]: True

await c.connect()
c.uaclient.protocol
Out[]: <asyncua.client.ua_client.UASocketProtocol at 0x7fe0870ec520>

c.uaclient.protocol.state
Out[]: 'open'

await c.disconnect()
c.uaclient.protocol.state
Out[]: 'closed'
```
So a condition like this should do it
```
if c.uaclient.protocol == None or c.uaclient.protocol.state == "closed":
    print("its closed")
```

**Comment by TheWicklowWolf** [comment, marked as answer]
Who wouldn't like some Wicklow Wolf beers?

The sync library doesn't have `uaclient`

I think `conn.tloop.is_alive()` will work for me...

**Reply by ZuZuD** [reply]
Oh yes, I didn't catch that. Out of curiosity, why do you favour the sync over the async client?

**Reply by TheWicklowWolf** [reply]
Mainly because I need to get the data at a defined interval...

**Reply by oroulet** [reply]
That is not a reason to use the sync api?!

**Reply by ZuZuD** [reply]
Yes, you can certainly achieve the same with the async client. I can't think of any use-case that would warrant using the sync client nowadays, outside maybe of having to run python < 3.5

**Reply by oroulet** [reply]
a good reason to use the sync client is if it has to be used inside a sync application...that is why we use it..

**Reply by oroulet** [reply]
in reality we are using the sync wrapper in the sync part of an async application . Quite horrible but no time to rewrite things...

**Reply by ZuZuD** [reply]
> a good reason to use the sync client is if it has to be used inside a sync application...that is why we use it..

You can wrap all your async code with `asyncio.run()` quite easily these days.

**Reply by dmitry-av** [reply]
Sync client actually has uaclient, it can be accessed via 
`client.aio_obj.uaclient.protocol.state`
tloop.is_alive() returns True even if socket connection was closed for some reason
